### ISSUE TITLE
alt-svc cache file grows unbounded

### ISSUE THREAD

**Issue by candrews** [curl/curl#5552]
### I did this
I'm running Kodi 19 with curl's alt-svc cache enabled, see https://github.com/xbmc/xbmc/pull/16618
I noticed that rarely, the alt-svc cache file will suddenly explode in size, growing to gigabytes in seconds. When this happens, the file's contents are:
```
# Your alt-svc cache. https://curl.haxx.se/docs/alt-svc.html
# This file waus generated by libcurl! Edit at your own risk.
h2 i.ytimg.com 443 h3-27 i.ytimg.com 443 "20200625 00:29:25" 0 0
h2 i.ytimg.com 443 h3-27 i.ytimg.com 443 "20200625 00:29:25" 0 0
h2 i.ytimg.com 443 h3-27 i.ytimg.com 443 "20200625 00:29:25" 0 0
h2 i.ytimg.com 443 h3-27 i.ytimg.com 443 "20200625 00:29:25" 0 0
h2 i.ytimg.com 443 h3-27 i.ytimg.com 443 "20200625 00:29:25" 0 0
h2 i.ytimg.com 443 h3-27 i.ytimg.com 443 "20200625 00:29:25" 0 0
```
The last line is repeated _many_ times. The time is even the same. This obviously shouldn't happen :)

Unfortunately, I don't have a minimum reproducible test case. I'm hoping that perhaps the problem is obvious to someone familiar with curl's internals.

FWIW, I tried to reproduce the issue using GNU Parallel:
`parallel -n0 'curl --alt-svc altcache.txt --silent --output /dev/null https://i.ytimg.com' ::: {0..99}`

After that, `altcache.txt` contains:
```
# Your alt-svc cache. https://curl.haxx.se/docs/alt-svc.html
# This file was generated by libcurl! Edit at your own risk.
h2 i.ytimg.com 443 h3-27 i.ytimg.com 443 "20200711 01:32:22" 0 0
h1 i.ytimg.com 443 h3-27 i.ytimg.com 443 "20200711 01:32:23" 0 0
```
which seems okay.

Maybe the issue has something to do with the connection cache or another feature not easily reproducible using the command line but that does happen in Kodi?

### I expected the following
The alt-svc cache file should not grow unbounded with the same host specified repeatedly.

### curl/libcurl version
```
$ curl -V
curl 7.70.0 (x86_64-pc-linux-gnu) libcurl/7.70.0 OpenSSL/1.1.1g zlib/1.2.11 brotli/1.0.7 libidn2/2.3.0 libssh2/1.9.0_DEV nghttp2/1.41.0 quiche/0.4.0 librtmp/2.3
Release-Date: 2020-04-29
Protocols: dict file ftp ftps http https imap imaps pop3 pop3s rtmp rtsp scp sftp smtp smtps tftp 
Features: alt-svc AsynchDNS brotli HTTP2 HTTP3 HTTPS-proxy IDN IPv6 Largefile libz Metalink NTLM SSL TLS-SRP UnixSockets
```
### operating system

Gentoo/Linux
```
$ uname -a
Linux irrational 5.4.33-gentoo-x86_64 #2 SMP Sat Apr 18 15:49:12 EDT 2020 x86_64 Intel(R) Core(TM) i5-6500 CPU @ 3.20GHz GenuineIntel GNU/Linux
```

**Comment by jay** [comment]
https://github.com/curl/curl/blob/53cdc2c963e33bc0cc1a51ad2df79396202e07f8/lib/altsvc.c#L355-L364

altsvc_out there's nothing unusual it writes a single line Do you know if libcurl finished writing the file? Did you notice any increase in memory usage?

**Comment by bagder** [comment]
> gigabytes in seconds

That does more reek of memory corruption (like the linked list pointing to itself or something) rather than just missing de-duplication in the code if it went that fast and wasn't tied to the HTTP request frequency...

**Comment by candrews** [comment]
> Do you know if libcurl finished writing the file? Did you notice any increase in memory usage?

Memory use increases very rapidly, quickly exhausting system memory and the OOM killer terminates the process.

**Comment by roz3x** [comment]
as the same value is repeated many times , this might means something like a self pointer situation  ?? in that case , a temperory fix would be
```    
  if (n == e)
         break;
```

**Comment by sterchelen** [comment]
Hey @candrews, do you have more information/update to add?

**Comment by candrews** [comment]
> do you have more information/update to add?

Not really... I don't know to further track this down. If someone can share with me how to do so, I can provide more information.

I can say that this issue isn't a one time thing; it just happened again.
```bash
$ du -hs curl-alt-svc-cache.txt
3.8G    curl-alt-svc-cache.txt
```
The file size is suspiciously close to 4 GB.

The file is filled exclusively with this sequence, repeated over and over:
```
h1 r6---sn-ab5szn7s.googlevideo.com 443 h3-27 r6---sn-ab5szn7s.googlevideo.com 443 "20200715 14:40:16" 0 0
h2 i.ytimg.com 443 h3-27 i.ytimg.com 443 "20200715 14:39:11" 0 0
h1 i.ytimg.com 443 h3-27 i.ytimg.com 443 "20200715 14:39:11" 0 0
h1 r1---sn-vgqsrned.googlevideo.com 443 h3-27 r1---sn-vgqsrned.googlevideo.com 443 "20200715 14:40:16" 0 0
h2 i.ytimg.com 443 h3-27 i.ytimg.com 443 "20200715 14:39:11" 0 0
h1 i.ytimg.com 443 h3-27 i.ytimg.com 443 "20200715 14:39:11" 0 0
h1 r6---sn-ab5l6n6s.googlevideo.com 443 h3-27 r6---sn-ab5l6n6s.googlevideo.com 443 "20200715 14:39:08" 0 0
h1 r1---sn-vgqsrned.googlevideo.com 443 h3-27 r1---sn-vgqsrned.googlevideo.com 443 "20200715 14:39:08" 0 0
h2 i.ytimg.com 443 h3-27 i.ytimg.com 443 "20200715 14:39:11" 0 0
h1 i.ytimg.com 443 h3-27 i.ytimg.com 443 "20200715 14:39:11" 0 0
h2 i.ytimg.com 443 h3-27 i.ytimg.com 443 "20200715 14:39:11" 0 0
h1 i.ytimg.com 443 h3-27 i.ytimg.com 443 "20200715 14:39:11" 0 0
h2 i.ytimg.com 443 h3-27 i.ytimg.com 443 "20200715 14:39:11" 0 0
h1 i.ytimg.com 443 h3-27 i.ytimg.com 443 "20200715 14:39:11" 0 0
h2 i.ytimg.com 443 h3-27 i.ytimg.com 443 "20200715 14:39:11" 0 0
h1 i.ytimg.com 443 h3-27 i.ytimg.com 443 "20200715 14:39:11" 0 0
```

**Comment by jay** [comment]
I looked at this again a few days ago, and I still don't see how it could happen.

**Comment by candrews** [comment]
Could it be a problem in Kodi's side? If so, do you have any hints as to what to look for in the code there?

**Comment by jay** [comment]
Run it with valgrind. It will be slower. If there's a bug in the list handling I'm not seeing it. The altsvc list is modified using only the [llist](https://github.com/curl/curl/blob/curl-7_71_1/lib/llist.c) functions which are used elsewhere in libcurl.

**Comment by bagder** [comment]
I would use [@roz3x's suggestion](https://github.com/curl/curl/issues/5552#issuecomment-643760970) and insert some debug code that you can break-point on and investigate what happened when the problem has triggered.

Since nobody else seems able to reproduce this, we rely on more info from you @candrews!

**Comment by bagder** [comment]
@candrews I'm going to suggest we close this issue until we get more info. We can't reproduce and nothing seems to happen with it. I don't see any use in just keeping it open here. Do you object?

**Comment by StefanYohansson** [comment]
well, I will clone xbmc repo with @candrews branch and give it a try with Valgrind. I will attach my investigation output later today, maybe it helps you guys to find an answer but as far I can see it looks like an xbmc problem.

**Comment by StefanYohansson** [comment]
@bagder 

After some tries to compile xbmc and reproduce the problem, I am not able to reproduce yet. Maybe we can close this issue here and continue discuss this on xbmc pull request.

@candrews

I am trying to reproduce using a Pictures menu > add pictures > add http picture source > put my nginx with alt-svc > start a alternative webservice both pointing to my pictures folder. I dont know if the problem is related to the tor thing that you mentioned in the xbmc PR.

Basically you can run `cmake -DVERBOSE=TRUE ..` and `cmake -DENABLE_EVENTCLIENTS=TRUE ..` on kodi build folder and run cmake --build again. So you can try to run kodi with `valgrind ./kodi-x11 > log.txt 2>&1` and send us that log file.

**Comment by bagder** [comment]
Yes thanks, seems totally reasonable. If you figure out any details that might hint on this being a curl problem we can always reopen or create a new issue!
